pysilcam/tracking/track.py

```diff
@@ -0,0 +1,417 @@
+        plt.title(str(timestamp))
+
+        subset = data[data['t2'] <= pd.to_datetime(timestamp)]
+
+        for p in subset['ParticleName'].values:
+            this_particle = subset[subset['ParticleName'] == p]
+            if len(this_particle) == 0:
+                continue
+
+            t_arr = min(this_particle['t1'])
+            t_dep = max(this_particle['t2'])
+            x_arrival = this_particle[this_particle['t1'] == t_arr]['x1'].values[0]
+            x_departure = this_particle[this_particle['t2'] == t_dep]['x2'].values[0]
+            linex = np.float64([x_arrival, x_departure])
+            y_arrival = this_particle[this_particle['t1'] == t_arr]['y1'].values[0]
+            y_departure = this_particle[this_particle['t2'] == t_dep]['y2'].values[0]
+            liney = np.float64([y_arrival, y_departure])
+
+            speed = calculate_speed(x_arrival, x_departure, y_arrival, y_departure, t_arr, t_dep, PIX_SIZE)
+
+            linex *= PIX_SIZE / 1000
+            liney *= PIX_SIZE / 1000
+            linex = c * PIX_SIZE / 1000 - linex
+            liney = r * PIX_SIZE / 1000 - liney
+            n_tracks = len(this_particle)  # should be the same as max(this_particle['n-tracks'])
+            plot_color = 'r-'
+            if n_tracks > track_length_limit:
+                plot_color = 'g-'
+
+                plt.text(liney[0], linex[0],
+                         (p + '\n{:0.2f}mm {:0.2f}mm/s'.format(this_particle['length'].values[0] * PIX_SIZE / 1000,
+                                                               speed * 10)),
+                         fontsize=4, color='g')
+
+            plt.plot(liney, linex, plot_color, linewidth=1)
+
+        plt.xlabel('mm')
+        plt.ylabel('mm')
+        plt.gca().invert_yaxis()
+
+        name = pd.to_datetime(timestamp).strftime('D%Y%m%dT%H%M%S.%f')
+        print('saving', os.path.join(outputdir, name + '-tr.png'))
+        plt.savefig(os.path.join(outputdir, name + '-tr.png'), dpi=300, bbox_inches='tight')
+
+
+def calculate_speed(x_arr, x_dep, y_arr, y_dep, t_arr, t_dep, PIX_SIZE):
+    X = np.array([x_arr, x_dep])
+    Y = np.array([y_arr, y_dep])
+
+    dt = pd.Series(np.abs(t_arr - t_dep)).dt.total_seconds()
+    dt = dt.values
+
+    Y_mm = Y * PIX_SIZE * 1e-3
+    X_mm = X * PIX_SIZE * 1e-3
+    dY_mm = np.diff(Y_mm, axis=0)
+    dX_mm = np.diff(X_mm, axis=0)
+    dD_mm = np.sqrt(dY_mm ** 2 + dX_mm ** 2)
+    dD_m = dD_mm / 1000
+    S_ms = dD_m / dt  # speed in m/s
+    S_cms = S_ms * 100  # speed in cm/s
+
+    return S_cms[0]
+
+
+def checkgroup(h5filename, groupstr):
+    '''check if a groupstr exists in the hdf5 file, h5filename
+    Args:
+        h5filename (str)        : hdf5 file name
+        groupstr (str)          : path to search for existence
+
+    Returns:
+        bool
+    '''
+    objs = []
+    with h5py.File(h5filename) as f:
+        f.visit(objs.append)
+        groups = [obj for obj in objs if isinstance(f[obj], h5py.Group)]
```

Comment:
F405: 'h5py' may be undefined, or defined from star imports: pysilcam.tracking.silcamtracker  <!-- Automatically posted by Lintly -->